File: Autonomous_Mobile_Robots_Course/Chapter12/Lesson3/Chapter12_Lesson3.cpp
// Chapter12_Lesson3.cpp
// Pose Graph Optimization in SE(2) via Gauss-Newton (Eigen, sparse)
// Build (Linux/Mac):
//   g++ -O2 -std=c++17 Chapter12_Lesson3.cpp -I /path/to/eigen -o pg
// Run:
//   ./pg
//
// Notes:
// - This is an educational from-scratch optimizer (no Ceres/g2o).
// - For real systems, use Ceres Solver, g2o, or GTSAM.

#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <Eigen/SparseCholesky>
#include <iostream>
#include <vector>
#include <cmath>

struct EdgeSE2 {
  int i;
  int j;
  Eigen::Vector3d z;      // [dx, dy, dtheta] measurement (i->j in i-frame)
  Eigen::Matrix3d Omega;  // information
};

static inline double wrap_angle(double a) {
  a = std::fmod(a + M_PI, 2.0 * M_PI);
  if (a < 0) a += 2.0 * M_PI;
  return a - M_PI;
}

static inline Eigen::Matrix2d rot2(double th) {
  double c = std::cos(th), s = std::sin(th);
  Eigen::Matrix2d R;
  R << c, -s,
       s,  c;
  return R;
}

// predict zhat = inv(xi) ⊕ xj
static inline Eigen::Vector3d predict_relative(const Eigen::Vector3d& xi,
                                               const Eigen::Vector3d& xj) {
  Eigen::Vector2d ti = xi.head<2>();
  Eigen::Vector2d tj = xj.head<2>();
  Eigen::Matrix2d Ri = rot2(xi(2));
  Eigen::Vector2d dt = Ri.transpose() * (tj - ti);
  double dth = wrap_angle(xj(2) - xi(2));
  return Eigen::Vector3d(dt(0), dt(1), dth);
}

// e = inv(z) ⊕ zhat (minimal translation+angle approximation)
static inline Eigen::Vector3d se2_error(const Eigen::Vector3d& xi,
                                        const Eigen::Vector3d& xj,
                                        const Eigen::Vector3d& z) {
  Eigen::Vector3d zhat = predict_relative(xi, xj);
  Eigen::Matrix2d Rz = rot2(z(2));
  Eigen::Vector2d terr = Rz.transpose() * (zhat.head<2>() - z.head<2>());
  double therr = wrap_angle(zhat(2) - z(2));
  return Eigen::Vector3d(terr(0), terr(1), therr);
}

static inline void se2_jacobians(const Eigen::Vector3d& xi,
                                 const Eigen::Vector3d& xj,
                                 const Eigen::Vector3d& z,
                                 Eigen::Matrix3d& A,
                                 Eigen::Matrix3d& B) {
  Eigen::Vector2d ti = xi.head<2>();
  Eigen::Vector2d tj = xj.head<2>();
  Eigen::Matrix2d Ri = rot2(xi(2));
  Eigen::Matrix2d Rz = rot2(z(2));
  Eigen::Matrix2d A2 = Rz.transpose() * Ri.transpose();

  Eigen::Matrix2d S;
  S << 0.0, -1.0,
       1.0,  0.0;

  Eigen::Vector2d dt = (tj - ti);

  A.setZero();
  B.setZero();

  A.block<2,2>(0,0) = -A2;
  B.block<2,2>(0,0) =  A2;

  Eigen::Vector2d dti_dtheta = Rz.transpose() * (-Ri.transpose() * (S * dt));
  A(0,2) = dti_dtheta(0);
  A(1,2) = dti_dtheta(1);

  A(2,2) = -1.0;
  B(2,2) =  1.0;
}

static inline void apply_increment(std::vector<Eigen::Vector3d>& X,
                                   const Eigen::VectorXd& dx,
                                   bool anchor_first_pose = true) {
  int N = (int)X.size();
  if (anchor_first_pose) {
    for (int k = 1; k < N; ++k) {
      Eigen::Vector3d d = dx.segment<3>(3*(k-1));
      X[k](0) += d(0);
      X[k](1) += d(1);
      X[k](2) = wrap_angle(X[k](2) + d(2));
    }
  } else {
    for (int k = 0; k < N; ++k) {
      Eigen::Vector3d d = dx.segment<3>(3*k);
      X[k](0) += d(0);
      X[k](1) += d(1);
      X[k](2) = wrap_angle(X[k](2) + d(2));
    }
  }
}

static inline void build_normal_equations(const std::vector<Eigen::Vector3d>& X,
                                          const std::vector<EdgeSE2>& edges,
                                          Eigen::SparseMatrix<double>& H,
                                          Eigen::VectorXd& b,
                                          double& chi2,
                                          bool anchor_first_pose = true) {
  int N = (int)X.size();
  int dim = anchor_first_pose ? 3*(N-1) : 3*N;

  std::vector<Eigen::Triplet<double>> trips;
  trips.reserve(edges.size() * 36); // rough

  b.setZero(dim);
  chi2 = 0.0;

  auto idx = [&](int pose_id) {
    return anchor_first_pose ? 3*(pose_id - 1) : 3*pose_id;
  };

  for (const auto& e : edges) {
    const auto& xi = X[e.i];
    const auto& xj = X[e.j];

    Eigen::Vector3d r = se2_error(xi, xj, e.z);
    Eigen::Matrix3d A, Bm;
    se2_jacobians(xi, xj, e.z, A, Bm);

    chi2 += r.transpose() * e.Omega * r;

    Eigen::Matrix3d AtOmega = A.transpose() * e.Omega;
    Eigen::Matrix3d BtOmega = Bm.transpose() * e.Omega;

    Eigen::Matrix3d Hii = AtOmega * A;
    Eigen::Matrix3d Hij = AtOmega * Bm;
    Eigen::Matrix3d Hji = BtOmega * A;
    Eigen::Matrix3d Hjj = BtOmega * Bm;

    Eigen::Vector3d bi = AtOmega * r;
    Eigen::Vector3d bj = BtOmega * r;

    auto addBlock = [&](int p, int q, const Eigen::Matrix3d& M) {
      if (anchor_first_pose && (p == 0 || q == 0)) return;
      int ip = idx(p), iq = idx(q);
      for (int a = 0; a < 3; ++a)
        for (int c = 0; c < 3; ++c)
          trips.emplace_back(ip + a, iq + c, M(a,c));
    };

    addBlock(e.i, e.i, Hii);
    addBlock(e.i, e.j, Hij);
    addBlock(e.j, e.i, Hji);
    addBlock(e.j, e.j, Hjj);

    if (!(anchor_first_pose && e.i == 0)) {
      int ii = idx(e.i);
      b.segment<3>(ii) += bi;
    }
    if (!(anchor_first_pose && e.j == 0)) {
      int jj = idx(e.j);
      b.segment<3>(jj) += bj;
    }
  }

  H.resize(dim, dim);
  H.setFromTriplets(trips.begin(), trips.end());
}

static inline void make_synthetic_pose_graph(int N,
                                             std::vector<Eigen::Vector3d>& X0,
                                             std::vector<EdgeSE2>& edges) {
  std::vector<Eigen::Vector3d> Xtrue(N);
  Xtrue[0].setZero();

  for (int k = 1; k < N; ++k) {
    Xtrue[k](0) = Xtrue[k-1](0) + 0.5 * std::cos(0.1 * k);
    Xtrue[k](1) = Xtrue[k-1](1) + 0.5 * std::sin(0.1 * k);
    Xtrue[k](2) = wrap_angle(0.05 * k);
  }

  double noise_xy = 0.02, noise_th = 0.01;
  Eigen::Matrix3d Sigma = Eigen::Matrix3d::Zero();
  Sigma(0,0) = noise_xy * noise_xy;
  Sigma(1,1) = noise_xy * noise_xy;
  Sigma(2,2) = noise_th * noise_th;

  Eigen::Matrix3d Omega = Sigma.inverse();

  auto randn = []() {
    // crude normal via Box-Muller
    static unsigned long long seed = 7ULL;
    seed = 6364136223846793005ULL * seed + 1ULL;
    double u1 = ((seed >> 11) * (1.0 / 9007199254740992.0));
    seed = 6364136223846793005ULL * seed + 1ULL;
    double u2 = ((seed >> 11) * (1.0 / 9007199254740992.0));
    double r = std::sqrt(-2.0 * std::log(std::max(1e-12, u1)));
    double th = 2.0 * M_PI * u2;
    return r * std::cos(th);
  };

  edges.clear();
  edges.reserve(N);

  for (int k = 0; k < N-1; ++k) {
    Eigen::Vector3d z = predict_relative(Xtrue[k], Xtrue[k+1]);
    z(0) += noise_xy * randn();
    z(1) += noise_xy * randn();
    z(2) = wrap_angle(z(2) + noise_th * randn());
    edges.push_back({k, k+1, z, Omega});
  }

  // loop closure 0 -> N-1
  Eigen::Vector3d zlc = predict_relative(Xtrue[0], Xtrue[N-1]);
  zlc(0) += noise_xy * randn();
  zlc(1) += noise_xy * randn();
  zlc(2) = wrap_angle(zlc(2) + noise_th * randn());
  edges.push_back({0, N-1, zlc, Omega});

  // initial guess: integrate noisy odometry only
  X0.assign(N, Eigen::Vector3d::Zero());
  for (int k = 0; k < N-1; ++k) {
    Eigen::Vector3d z = edges[k].z;
    Eigen::Matrix2d Rk = rot2(X0[k](2));
    X0[k+1].head<2>() = X0[k].head<2>() + Rk * z.head<2>();
    X0[k+1](2) = wrap_angle(X0[k](2) + z(2));
  }
}

int main() {
  std::vector<Eigen::Vector3d> X;
  std::vector<EdgeSE2> edges;

  make_synthetic_pose_graph(30, X, edges);

  std::cout << "Optimizing pose graph (Gauss-Newton, anchor pose 0)\n";

  int iters = 15;
  for (int it = 0; it < iters; ++it) {
    Eigen::SparseMatrix<double> H;
    Eigen::VectorXd b;
    double chi2 = 0.0;

    int dim = 3 * ((int)X.size() - 1);
    b.resize(dim);

    build_normal_equations(X, edges, H, b, chi2, true);

    Eigen::SimplicialLDLT<Eigen::SparseMatrix<double>> solver;
    solver.compute(H);
    if (solver.info() != Eigen::Success) {
      std::cerr << "Factorization failed\n";
      return 1;
    }
    Eigen::VectorXd dx = solver.solve(-b);

    apply_increment(X, dx, true);

    std::cout << "iter=" << it << "  chi2=" << chi2
              << "  |dx|=" << dx.norm() << "\n";

    if (dx.norm() < 1e-8) break;
  }

  std::cout << "\nFirst 5 poses (x,y,theta):\n";
  for (int k = 0; k < 5; ++k) {
    std::cout << "k=" << k << "  " << X[k].transpose() << "\n";
  }
  return 0;
}

File: Autonomous_Mobile_Robots_Course/Chapter12/Lesson3/Chapter12_Lesson3.py
#!/usr/bin/env python3
# Chapter12_Lesson3.py
# Pose Graph Optimization in SE(2) via Gauss-Newton (from scratch, educational)
#
# Dependencies:
#   numpy
#   scipy  (sparse linear algebra)
#
# Run:
#   python Chapter12_Lesson3.py

import numpy as np
from dataclasses import dataclass
from typing import List, Tuple
from scipy.sparse import lil_matrix, csr_matrix
from scipy.sparse.linalg import spsolve

def wrap_angle(a: float) -> float:
    """Wrap angle to (-pi, pi]."""
    return (a + np.pi) % (2.0 * np.pi) - np.pi

def rot2(theta: float) -> np.ndarray:
    c, s = np.cos(theta), np.sin(theta)
    return np.array([[c, -s],
                     [s,  c]], dtype=float)

@dataclass
class EdgeSE2:
    i: int
    j: int
    z: np.ndarray      # measurement [dx, dy, dtheta] from i to j (in i-frame)
    Omega: np.ndarray  # information matrix (3x3), inverse of covariance

def predict_relative(xi: np.ndarray, xj: np.ndarray) -> np.ndarray:
    """
    Predicted relative pose from i to j:
      zhat = inv(xi) ⊕ xj
    where pose x = [x, y, theta].
    """
    ti = xi[:2]
    tj = xj[:2]
    Ri = rot2(xi[2])
    dt = Ri.T @ (tj - ti)
    dtheta = wrap_angle(xj[2] - xi[2])
    return np.array([dt[0], dt[1], dtheta], dtype=float)

def se2_error(xi: np.ndarray, xj: np.ndarray, z: np.ndarray) -> np.ndarray:
    """
    Simple SE(2) 'between' error:
      e = inv(z) ⊕ (inv(xi) ⊕ xj)
    implemented using a minimal (translation + angle) approximation.
    """
    zhat = predict_relative(xi, xj)
    Rz = rot2(z[2])
    terr = Rz.T @ (zhat[:2] - z[:2])
    therr = wrap_angle(zhat[2] - z[2])
    return np.array([terr[0], terr[1], therr], dtype=float)

def se2_jacobians(xi: np.ndarray, xj: np.ndarray, z: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
    """
    Analytic Jacobians for the simple between error used in se2_error.

    e_t = Rz^T ( Ri^T (t_j - t_i) - t_z )
    e_th = (theta_j - theta_i) - theta_z  (wrapped)

    Returns (A, B) where:
      A = de/dxi  (3x3), B = de/dxj  (3x3)
    """
    ti = xi[:2]
    tj = xj[:2]
    Ri = rot2(xi[2])
    Rz = rot2(z[2])

    # A2 = Rz^T Ri^T
    A2 = Rz.T @ Ri.T

    # skew-symmetric matrix for 2D cross product
    S = np.array([[0.0, -1.0],
                  [1.0,  0.0]])

    dt = (tj - ti)

    A = np.zeros((3, 3), dtype=float)
    B = np.zeros((3, 3), dtype=float)

    # translation derivatives
    A[0:2, 0:2] = -A2
    B[0:2, 0:2] =  A2

    # derivative wrt theta_i affects translation part via Ri^T
    # d(Ri^T dt)/dtheta_i = -Ri^T S dt
    dti_dtheta = Rz.T @ (-Ri.T @ (S @ dt))
    A[0:2, 2] = dti_dtheta

    # theta derivatives
    A[2, 2] = -1.0
    B[2, 2] =  1.0

    return A, B

def build_normal_equations(
    X: np.ndarray,
    edges: List[EdgeSE2],
    anchor_first_pose: bool = True
) -> Tuple[csr_matrix, np.ndarray, float]:
    """
    Build sparse normal equations H dx = -b for Gauss-Newton.

    If anchor_first_pose=True, pose 0 is held fixed (gauge fixing) and variables
    are only for poses 1..N-1.
    """
    N = X.shape[0]
    dim = 3 * (N - 1) if anchor_first_pose else 3 * N

    H = lil_matrix((dim, dim), dtype=float)
    b = np.zeros(dim, dtype=float)
    chi2 = 0.0

    def idx(pose_id: int) -> int:
        if anchor_first_pose:
            return 3 * (pose_id - 1)  # pose 1 -> 0
        return 3 * pose_id

    for e in edges:
        xi = X[e.i]
        xj = X[e.j]
        r = se2_error(xi, xj, e.z)
        A, B = se2_jacobians(xi, xj, e.z)

        Omega = e.Omega
        chi2 += float(r.T @ Omega @ r)

        # local blocks
        AtOmega = A.T @ Omega
        BtOmega = B.T @ Omega
        Hii = AtOmega @ A
        Hij = AtOmega @ B
        Hji = BtOmega @ A
        Hjj = BtOmega @ B
        bi = AtOmega @ r
        bj = BtOmega @ r

        # scatter-add into global H and b, skipping anchored pose if needed
        for (p, q, Hpq) in [(e.i, e.i, Hii), (e.i, e.j, Hij), (e.j, e.i, Hji), (e.j, e.j, Hjj)]:
            if anchor_first_pose and (p == 0 or q == 0):
                continue
            ip = idx(p)
            iq = idx(q)
            for a in range(3):
                for bb in range(3):
                    H[ip + a, iq + bb] += Hpq[a, bb]

        # b terms
        if not (anchor_first_pose and e.i == 0):
            ii = idx(e.i)
            b[ii:ii+3] += bi
        if not (anchor_first_pose and e.j == 0):
            jj = idx(e.j)
            b[jj:jj+3] += bj

    return H.tocsr(), b, chi2

def apply_increment(X: np.ndarray, dx: np.ndarray, anchor_first_pose: bool = True) -> np.ndarray:
    """Apply increments to poses; theta updated additively and wrapped."""
    X_new = X.copy()
    if anchor_first_pose:
        for k in range(1, X.shape[0]):
            d = dx[3*(k-1):3*(k-1)+3]
            X_new[k, 0] += d[0]
            X_new[k, 1] += d[1]
            X_new[k, 2] = wrap_angle(X_new[k, 2] + d[2])
    else:
        for k in range(X.shape[0]):
            d = dx[3*k:3*k+3]
            X_new[k, 0] += d[0]
            X_new[k, 1] += d[1]
            X_new[k, 2] = wrap_angle(X_new[k, 2] + d[2])
    return X_new

def gauss_newton(
    X0: np.ndarray,
    edges: List[EdgeSE2],
    iters: int = 10,
    damping: float = 0.0,
    anchor_first_pose: bool = True
) -> np.ndarray:
    X = X0.copy()
    for it in range(iters):
        H, b, chi2 = build_normal_equations(X, edges, anchor_first_pose=anchor_first_pose)
        if damping > 0.0:
            H = H + damping * csr_matrix(np.eye(H.shape[0]))

        dx = spsolve(H, -b)
        X = apply_increment(X, dx, anchor_first_pose=anchor_first_pose)

        print(f"iter={it:02d}  chi2={chi2:.6f}  |dx|={np.linalg.norm(dx):.3e}")

        if np.linalg.norm(dx) < 1e-8:
            break
    return X

def make_synthetic_pose_graph(N: int = 20, noise_xy: float = 0.02, noise_th: float = 0.01):
    """
    Build a simple chain with one loop-closure edge.
    Ground truth: gentle arc.
    Measurements: relative odometry edges (i,i+1) + loop edge (0,N-1).
    """
    X_true = np.zeros((N, 3), dtype=float)
    for k in range(1, N):
        X_true[k, 0] = X_true[k-1, 0] + 0.5 * np.cos(0.1 * k)
        X_true[k, 1] = X_true[k-1, 1] + 0.5 * np.sin(0.1 * k)
        X_true[k, 2] = wrap_angle(0.05 * k)

    edges = []
    Sigma = np.diag([noise_xy**2, noise_xy**2, noise_th**2])
    Omega = np.linalg.inv(Sigma)

    rng = np.random.default_rng(7)

    for k in range(N-1):
        z = predict_relative(X_true[k], X_true[k+1])
        z_noisy = z + rng.normal(0.0, [noise_xy, noise_xy, noise_th])
        z_noisy[2] = wrap_angle(z_noisy[2])
        edges.append(EdgeSE2(i=k, j=k+1, z=z_noisy, Omega=Omega))

    zlc = predict_relative(X_true[0], X_true[N-1])
    zlc_noisy = zlc + rng.normal(0.0, [noise_xy, noise_xy, noise_th])
    zlc_noisy[2] = wrap_angle(zlc_noisy[2])
    edges.append(EdgeSE2(i=0, j=N-1, z=zlc_noisy, Omega=Omega))

    X0 = np.zeros_like(X_true)
    for k in range(N-1):
        z = edges[k].z
        Rk = rot2(X0[k,2])
        X0[k+1,:2] = X0[k,:2] + (Rk @ z[:2])
        X0[k+1,2] = wrap_angle(X0[k,2] + z[2])

    return X0, edges

def main():
    X0, edges = make_synthetic_pose_graph(N=30)
    print("Optimizing pose graph (Gauss-Newton, anchored pose 0)...")
    X_opt = gauss_newton(X0, edges, iters=15, damping=0.0, anchor_first_pose=True)

    print("\nFirst 5 poses (x,y,theta) initial vs optimized:")
    for k in range(5):
        print(f"k={k:02d}  X0={X0[k]}  Xopt={X_opt[k]}")

if __name__ == "__main__":
    main()
